securityscanutils/github_writer.go

```diff
@@ -0,0 +1,127 @@
+	return fmt.Sprintf("github.com/%s/%s", r.Owner, r.RepoName)
+}
+
+// GithubIssueWriter is responsible for creating Github issues
+// to track vulnerabilities that have been discovered in images within a release.
+// It is configured with a Predicate that filters which releases to
+// write issues for, and which to skip
+type GithubIssueWriter struct {
+	// The details about the Github repository
+	repo GithubRepo
+
+	// The client used to write the issues
+	client *github.Client
+
+	// The RepositoryReleasePredicate used to determine if a vulnerability
+	// associated with a certain release should be uploaded to GitHub
+	createGithubIssuePredicate githubutils.RepositoryReleasePredicate
+
+	// A local cache of all existing GitHub issues
+	// Used to ensure that we are updating existing issues that were created by previous scans
+	allGithubIssues []*github.Issue
+}
+
+func NewGithubIssueWriter(repo GithubRepo, client *github.Client, issuePredicate githubutils.RepositoryReleasePredicate) *GithubIssueWriter {
+	return &GithubIssueWriter{
+		repo:                       repo,
+		client:                     client,
+		createGithubIssuePredicate: issuePredicate,
+		allGithubIssues:            nil, // initially nil, we'll lazy load these
+	}
+}
+
+// Labels that are applied to github issues that security scan generates
+var labels = []string{"trivy", "vulnerability"}
+
+// getAllGithubIssues returns the set of open issues in a Github repository that contain the trivy labels
+func (g *GithubIssueWriter) getAllGithubIssues(ctx context.Context) ([]*github.Issue, error) {
+	if g.allGithubIssues != nil {
+		// Maintain a local cache of issue to avoid re-requesting each time
+		return g.allGithubIssues, nil
+	}
+
+	issues, err := githubutils.GetAllIssues(ctx, g.client, g.repo.Owner, g.repo.RepoName, &github.IssueListByRepoOptions{
+		State:  "open",
+		Labels: labels,
+	})
+	if err != nil {
+		return nil, eris.Wrapf(err, "error fetching all issues from %s", g.repo.Address())
+	}
+	g.allGithubIssues = issues
+	return g.allGithubIssues, nil
+}
+
+// Creates/Updates a Github Issue per image
+// The github issue will have the markdown table report of the image's vulnerabilities
+// example: https://github.com/solo-io/solo-projects/issues/2458
+func (g *GithubIssueWriter) CreateUpdateVulnerabilityIssue(ctx context.Context, release *github.RepositoryRelease, vulnerabilityMarkdown string) error {
+	if vulnerabilityMarkdown == "" {
+		// There we no vulnerabilities discovered for this release
+		// do not create an empty github issue
+		return nil
+	}
+
+	if !g.shouldWriteIssue(release) {
+		// The GithubIssueWriter can be configured to only write issues for certain releases
+		return nil
+	}
+
+	// We can swallow the error here, any releases with improper tag names
+	// will not be included in the filtered list
+	versionToScan, _ := semver.NewVersion(release.GetTagName())
+
+	issueTitle := fmt.Sprintf("Security Alert: %s", versionToScan.String())
```

Comment:
nit: While we are changing things can we create our custom stringer
```suggestion
	issueTitle := fmt.Sprintf("Security Alert: %d.%d", versionToScan.Major(), versionToScan.Minor())
```

Reply:
Spoke offline:

The challenge is that since we support the feature where you can create an issue per version, its possible to have multiple issues with the same major.minor and we'd therefore have issues with overlapping names.

By keeping the major.minor.patch in the title, we ensure that we create distinct issues. Also, this allows us to close issues when we resolve them, since a new issue (with a new title) will be created for the next version with a vulnerability.